feat(docs): Add mining instructions to the zebra book (#6199)
ZcashFoundation/zebra: book/src/user/mining.md

# Mining Zcash with zebra

Zebra's RPC methods support miners and mining pools. 

Contents:

- [Download and build Zebra](#download-and-build-zebra)
- [Configure zebra for mining](#configure-zebra-for-mining)
  - [Miner address](#miner-address)
  - [RPC section](#rpc-section)
- [Running zebra](#running-zebra)
- [Testing the setup](#testing-the-setup)
- [Run a mining pool](#run-a-mining-pool)

## Download and build Zebra for mining
[#download-and-build-zebra]: #download-and-build-zebra

> 

Reviewer: Can you link to the README section that lists the Zebra dependencies that need to be installed?
That should be the first step before building.

If we link to the list of dependencies, we will only have to update them in one place.

```console
cargo install --locked --features getblocktemplate-rpcs --git https://github.com/ZcashFoundation/zebra zebrad
```

Zebra binary will be at ` ~/.cargo/bin/zebrad`.

It is important to install with the `getblocktemplate-rpcs` feature so the final binary will contain mining support.

## Configure zebra for mining
[#configure-zebra-for-mining]: #configure-zebra-for-mining

We need a configuration file. First, we create a file with the default settings:

```console

> zebrad generate -o ~/.config/zebrad.toml

Reviewer: If the directory does not exist, we need to create it first.

```suggestion
mkdir -p ~/.config
zebrad generate -o ~/.config/zebrad.toml
```

```

The above command places the generated `zebrad.toml` config file in the default preferences directory of Linux. For other OSes default locations [see here](https://docs.rs/dirs/latest/dirs/fn.preference_dir.html).

Tweak the following options in order to prepare for mining.

### Miner address
[#miner-address]: #miner-address

Node miner address is required. At the moment zebra only allows `p2pkh` or `p2sh` transparent addresses.

```
[mining]

> miner_address = 't3K4aLYagSSBySdrfAGGeUd5H9z5Qvz88t2'

Reviewer: Where is this address from? (Which network and who has the keys?)

This needs to go as a note in the docs, so we are transparent about who gets the funds from it.

```

### RPC section
[#rpc-section]: #rpc-section

> This change is required for zebra to behave as an RPC endpoint. The standard port for RPC endpoint is `8232`.

Reviewer: ```suggestion
This change is required for zebra to behave as an RPC endpoint. The standard port for RPC endpoint is `8232` on mainnet.
```

```
[rpc]
listen_addr = "127.0.0.1:8232"
```

## Running zebra
[#running-zebra]: #running-zebra

If the configuration file is in the default directory, then zebra will just read from it. All we need to do is to start zebra as follows:

```console
zebrad
```

You can specify the configuration file path with `-c /path/to/config.file`.

Wait until zebra is in sync, you will see the sync at 100% when this happens:

```console
...
2023-02-21T18:41:09.088931Z  INFO {zebrad="4daedbc" net="Main"}: zebrad::components::sync::progress: finished initial sync to chain tip, using gossiped blocks sync_percent=100.000% current_height=Height(1992055) network_upgrade=Nu5 remaining_sync_blocks=1 time_since_last_state_block=0s
...
```

## Testing the setup
[#testing-the-setup]: #testing-the-setup

> 

Reviewer: This large command output might make the docs hard to read. Can you put it in a details block? Or replace it with a smaller example?

https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/organizing-information-with-collapsed-sections

      "hash": "f77c29f032f4abe579faa891c8456602f848f423021db1f39578536742e8ff3e",
      "authdigest": "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
      "depends": [],
      "fee": -12852,
      "sigops": 0,
      "required": true
    },
    "longpollid": "0001992055c6e3ad7916770099070000000004516b4994",
    "target": "0000000001a11f00000000000000000000000000000000000000000000000000",
    "mintime": 1677004508,
    "mutable": [
      "time",
      "transactions",
      "prevblock"
    ],
    "noncerange": "00000000ffffffff",
    "sigoplimit": 20000,
    "sizelimit": 2000000,
    "curtime": 1677004885,
    "bits": "1c01a11f",
    "height": 1992056,
    "maxtime": 1677009907
  },
  "id": "curltest"
}
$ 
```

## Run a mining pool
[#run-a-mining-pool]: #run-a-mining-pool

> Just point your mining pool software to the Zebra RPC endpoint (127.0.0.1:8232). Zebra supports all the RPC methods needed to run the [s-nomp](https://github.com/ZcashFoundation/s-nomp) mining pool software.

Reviewer: I'm not sure if we should suggest that `s-nomp` is useful mining software?
It hasn't been updated for NU5, so I don't want to lead people towards it.

```suggestion
Just point your mining pool software to the Zebra RPC endpoint (127.0.0.1:8232). Zebra supports the RPC methods needed to run most mining pool software.
```

> If you want to run an `s-nomp` mining pool with Zebra please refer to [this document](https://github.com/ZcashFoundation/s-nomp/blob/zebra-mining/zebra-instructions.md) for a very detailed guide.

Reviewer: 1. Can we make it clear that our `s-nomp` fork is experimental, and why?
2. We are going to move the detailed doc to the Zebra repository, so I opened PR #6201 to move that file, and updated the link here

```suggestion
If you want to run an experimental `s-nomp` mining pool with Zebra on testnet, please refer to [this document](https://github.com/ZcashFoundation/zebra/blob/main/book/src/user/mining-testnet-s-nomp.md) for a very detailed guide. `s-nomp` is not compatible with NU5, so some mining functions are disabled.
```
